samples/subsys/modbus/tcp_server/src/main.c

```diff
@@ -213,14 +213,14 @@ static int init_modbus_server(void)
 
 static int modbus_tcp_reply(int client, struct modbus_adu *adu)
 {
-	uint8_t header[MODBUS_MBAP_AND_FC_LENGTH];
+	uint8_t message[MODBUS_MBAP_AND_FC_LENGTH + adu->length];
 
-	modbus_raw_put_header(adu, header);
-	if (send(client, header, sizeof(header), 0) < 0) {
-		return -errno;
-	}
+	modbus_raw_put_header(adu, message);
+
+	memcpy(&message[MODBUS_MBAP_AND_FC_LENGTH], adu->data,
+	       MIN(MODBUS_MBAP_AND_FC_LENGTH + adu->length, CONFIG_MODBUS_BUFFER_SIZE));
 
-	if (send(client, adu->data, adu->length, 0) < 0) {
+	if (send(client, message, sizeof(message), 0) < 0) {
```

Comment:
There is no guarantee in TCP that the packet is sent in one go, so it really is a peer issue if it cannot accept the data in separate reads. We can try to mitigate this by using `sendmsg()` call which can have separate buffers and which should (not guaranteed) be able to place the sent data to one TCP segment.

Reply:
I was working with an PLC/HMI vendor, and hoped that they'd change their system to accept the packets in separate reads, but was informed that it was non-standard behavior.
They told me to send it in one go, which got it to work. I will try looking into using `sendmsg()` call...

Reply:
It looks like the vendor wants to avoid fixing their system, they do not seem to understand how TCP works.